pkg/svc/workflow.go

```diff
@@ -2,173 +2,123 @@ package svc
 
 import (
 	"context"
-	"errors"
-	"fmt"
 	"log"
-	"time"
+	"strings"
 
 	"github.com/fuseml/fuseml-core/gen/workflow"
+	"github.com/fuseml/fuseml-core/pkg/core"
 	"github.com/fuseml/fuseml-core/pkg/core/config"
+	"github.com/fuseml/fuseml-core/pkg/core/manager"
 	"github.com/fuseml/fuseml-core/pkg/core/tekton"
 	"github.com/fuseml/fuseml-core/pkg/domain"
 )
 
-// createWorkflowListenerTimeout is the time (in minutes) that FuseML waits for the workflow listener
-// to be available
-const createWorkflowListenerTimeout = 1
-
 // workflow service example implementation.
 // The example methods log the requests and return zero values.
 type workflowsrvc struct {
-	logger       *log.Logger
-	store        domain.WorkflowStore
-	codesetStore domain.CodesetStore
-	backend      domain.WorkflowBackend
+	logger *log.Logger
+	mgr    domain.WorkflowManager
 }
 
 // NewWorkflowService returns the workflow service implementation.
-func NewWorkflowService(logger *log.Logger, store domain.WorkflowStore, codesetStore domain.CodesetStore) (workflow.Service, error) {
-	backend, err := tekton.NewWorkflowBackend(logger, config.FuseMLNamespace)
+// FIXME: instead of CodesetStore, receive CodesetManager
+func NewWorkflowService(logger *log.Logger, codesetStore domain.CodesetStore) (workflow.Service, error) {
+	workflowStore := core.NewWorkflowStore()
+	workflowBackend, err := tekton.NewWorkflowBackend(logger, config.FuseMLNamespace)
 	if err != nil {
+		logger.Print(err)
 		return nil, err
 	}
-	return &workflowsrvc{logger, store, codesetStore, backend}, nil
+	workflowMgr := manager.NewWorkflowManager(workflowBackend, workflowStore, codesetStore)
```

Comment:
Please move this initialization in main and pass the already initialized workflow manager as a parameter to this service. I mentioned this before: we need a centralized place where all these managers and their sub-components are initialized, because they will eventually need to reference one another, and right now main is the best choice. Spreading it out across REST API services doesn't make sense.

This also helps remove the tekton backend reference from here (yes, I know, it will land in main instead, but that's ok).

Reply:
I left it here on purpose 😄   and the reason is that we had this discussion before we did not reach an agreement.

As you already know, I do not agree on initialising the backend at `main` as this basically encourages any other service to consume it directly. IMO, the initialisation should be done by the workflow manager as it is the only object that should provide access to it through its exposed functions. Besides the workflow backend, I also think that the same can be applied to services store.

Reply:
Yeah, I already know you disagree :grin: , but IMHO you haven't brought any good arguments against it thus far.

> I do not agree on initialising the backend at main as this basically encourages any other service to consume it directly

This idea doesn't make sense to me. Services are not self aware, and for people there are code-review processes that prevent this from happening. 

> the initialisation should be done by the workflow manager as it is the only object that should provide access to it through its exposed functions

Did you mean to say the workflow service here ? The workflow manager shouldn't even know what a tekton backend is, given that it interacts with a generic backend interface. That's part of what makes this separation work: the dependency injection, or dependency inversion principle. But that's already done: the workflow manager doesn't depend on the tekton backend.

Doing the initializing in the REST service has some major disadvantages that I want to avoid:
- you have a dependency between REST services and backends, even though REST services only need to talk to the "managers"
- backends and managers are initialized in REST services instead of some neutral place
- managers are singletons and will most likely need to reference each-other. They will also need to be referenced from more than one REST service, so there's not a 1-to-1 relationship between REST endpoints/services and managers

I don't have a preference for using `main` to initialize all these components and use dependency injection to set up the way the use one another. We can use another file, or even another package, dedicated to initialization. As long as we do it on "neutral" ground: not in the managers, not in the REST services, not in other components.

Reply:
> Yeah, I already know you disagree 😁 , but IMHO you haven't brought any good arguments against it thus far.
> 
> > I do not agree on initialising the backend at main as this basically encourages any other service to consume it directly
> 
> This idea doesn't make sense to me. Services are not self aware, and for people there are code-review processes that prevent this from happening.

Why rely on code-review when the code structure itself can ensure that? Hoping that code-review will prevent any bad code to get in is a bit naive IMO.

> 
> > the initialisation should be done by the workflow manager as it is the only object that should provide access to it through its exposed functions
> 
> Did you mean to say the workflow service here ? The workflow manager shouldn't even know what a tekton backend is, given that it interacts with a generic backend interface. That's part of what makes this separation work: the dependency injection, or dependency inversion principle. But that's already done: the workflow manager doesn't depend on the tekton backend.
> 
> Doing the initializing in the REST service has some major disadvantages that I want to avoid:
> 
> * you have a dependency between REST services and backends, even though REST services only need to talk to the "managers"
> * backends and managers are initialized in REST services instead of some neutral place
> * managers are singletons and will most likely need to reference each-other. They will also need to be referenced from more than one REST service, so there's not a 1-to-1 relationship between REST endpoints/services and managers
> 
> I don't have a preference for using `main` to initialize all these components and use dependency injection to set up the way the use one another. We can use another file, or even another package, dedicated to initialization. As long as we do it on "neutral" ground: not in the managers, not in the REST services, not in other components.

I totally agree, manager indeed might not be the best place, but `main` also isn't. I guess I need a bit more of learning to figure out a better solution.

Reply:
Just for the record: we agreed to use main to do the initialization of components in this PR. Nonetheless, we also agreed that things will get more complicated as we add more components, especially when we need to have circular dependencies (such as the Workflow manager and Codeset manager depending on each-other to support removing codeset assignments automatically when a codeset is deleted). We'll need to evaluate better ways of maintaining the graph of component dependencies and doing dependency injection in the correct order, and a framework like wire could be the answer.